File: Einfach/SolutionsByDeepSeek/C++ Leaning/V1Code.py
import math
import random
from dataclasses import dataclass
from typing import List, Optional, Tuple
import struct

# ============================================================================
# Mathematik-Modul
# ============================================================================

@dataclass
class Vec3:
    """3D-Vektor mit grundlegenden Operationen"""
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0
    
    def __add__(self, other):
        return Vec3(self.x + other.x, self.y + other.y, self.z + other.z)
    
    def __sub__(self, other):
        return Vec3(self.x - other.x, self.y - other.y, self.z - other.z)
    
    def __mul__(self, scalar):
        return Vec3(self.x * scalar, self.y * scalar, self.z * scalar)
    
    def __rmul__(self, scalar):
        return self.__mul__(scalar)
    
    def __truediv__(self, scalar):
        return Vec3(self.x / scalar, self.y / scalar, self.z / scalar)
    
    def dot(self, other):
        return self.x * other.x + self.y * other.y + self.z * other.z
    
    def cross(self, other):
        return Vec3(
            self.y * other.z - self.z * other.y,
            self.z * other.x - self.x * other.z,
            self.x * other.y - self.y * other.x
        )
    
    def length(self):
        return math.sqrt(self.dot(self))
    
    def normalize(self):
        l = self.length()
        if l > 0:
            return self / l
        return self
    
    def __str__(self):
        return f"({self.x:.2f}, {self.y:.2f}, {self.z:.2f})"

@dataclass
class Ray:
    """Strahl mit Ursprung und Richtung"""
    origin: Vec3
    direction: Vec3
    
    def point_at(self, t: float) -> Vec3:
        return self.origin + self.direction * t

# ============================================================================
# Geometrie-Modul
# ============================================================================

class Material:
    """Material-Eigenschaften für Objekte"""
    def __init__(self, color: Vec3, emission: Vec3 = Vec3(), 
                 reflectance: float = 0.0, roughness: float = 0.0):
        self.color = color
        self.emission = emission
        self.reflectance = reflectance  # 0 = kein Spiegel, 1 = perfekter Spiegel
        self.roughness = roughness      # Oberflächenrauheit für diffuse Reflexion

class HitRecord:
    """Informationen über einen Trefferpunkt"""
    def __init__(self):
        self.t = float('inf')
        self.point = Vec3()
        self.normal = Vec3()
        self.material = None
        self.hit = False

class Sphere:
    """Kugel als geometrisches Primitive"""
    def __init__(self, center: Vec3, radius: float, material: Material):
        self.center = center
        self.radius = radius
        self.material = material
    
    def hit(self, ray: Ray, t_min: float, t_max: float) -> HitRecord:
        record = HitRecord()
        oc = ray.origin - self.center
        a = ray.direction.dot(ray.direction)
        b = oc.dot(ray.direction)
        c = oc.dot(oc) - self.radius * self.radius
        discriminant = b * b - a * c
        
        if discriminant > 0:
            sqrt_disc = math.sqrt(discriminant)
            t = (-b - sqrt_disc) / a
            if t < t_max and t > t_min:
                record.t = t
                record.point = ray.point_at(t)
                record.normal = (record.point - self.center) / self.radius
                record.material = self.material
                record.hit = True
                return record
            
            t = (-b + sqrt_disc) / a
            if t < t_max and t > t_min:
                record.t = t
                record.point = ray.point_at(t)
                record.normal = (record.point - self.center) / self.radius
                record.material = self.material
                record.hit = True
                return record
        
        return record

class Plane:
    """Ebene als geometrisches Primitive"""
    def __init__(self, point: Vec3, normal: Vec3, material: Material):
        self.point = point
        self.normal = normal.normalize()
        self.material = material
    
    def hit(self, ray: Ray, t_min: float, t_max: float) -> HitRecord:
        record = HitRecord()
        denom = self.normal.dot(ray.direction)
        
        if abs(denom) > 1e-6:
            t = (self.point - ray.origin).dot(self.normal) / denom
            if t < t_max and t > t_min:
                record.t = t
                record.point = ray.point_at(t)
                record.normal = self.normal
                record.material = self.material
                record.hit = True
        
        return record

class World:
    """Welt, die alle Objekte enthält"""
    def __init__(self):
        self.objects = []
    
    def add(self, obj):
        self.objects.append(obj)
    
    def hit(self, ray: Ray, t_min: float, t_max: float) -> HitRecord:
        closest_record = HitRecord()
        closest_record.t = t_max
        
        for obj in self.objects:
            record = obj.hit(ray, t_min, closest_record.t)
            if record.hit:
                closest_record = record
        
        return closest_record

# ============================================================================
# Raytracing-Modul
# ============================================================================

class Camera:
    """Kamera mit Position und Blickrichtung"""
    def __init__(self, lookfrom: Vec3, lookat: Vec3, vup: Vec3, 
                 vfov: float, aspect: float):
        self.origin = lookfrom
        self.lookat = lookat
        
        theta = vfov * math.pi / 180
        half_height = math.tan(theta / 2)
        half_width = aspect * half_height
        
        w = (lookfrom - lookat).normalize()
        u = vup.cross(w).normalize()
        v = w.cross(u)
        
        self.lower_left_corner = self.origin - u * half_width - v * half_height - w
        self.horizontal = u * (2 * half_width)
        self.vertical = v * (2 * half_height)
    
    def get_ray(self, u: float, v: float) -> Ray:
        direction = self.lower_left_corner + self.horizontal * u + self.vertical * v - self.origin
        return Ray(self.origin, direction.normalize())

class Raytracer:
    """Haupt-Raytracer mit Rekursion und Beleuchtungsberechnung"""
    def __init__(self, world: World, max_depth: int = 10):
        self.world = world
        self.max_depth = max_depth
        self.background_color = Vec3(0.1, 0.1, 0.1)
    
    def trace(self, ray: Ray, depth: int) -> Vec3:
        if depth >= self.max_depth:
            return Vec3()
        
        record = self.world.hit(ray, 0.001, float('inf'))
        
        if not record.hit:
            return self.background_color
        
        # Emission des getroffenen Materials
        color = record.material.emission
        
        # Reflexion berechnen
        if record.material.reflectance > 0:
            reflected = self.reflect(ray.direction, record.normal)
            # Rauheit durch zufällige Abweichung
            if record.material.roughness > 0:
                random_dir = Vec3(random.uniform(-1, 1), 
                                 random.uniform(-1, 1), 
                                 random.uniform(-1, 1)).normalize()
                reflected = (reflected + random_dir * record.material.roughness).normalize()
            
            reflected_ray = Ray(record.point, reflected)
            reflected_color = self.trace(reflected_ray, depth + 1)
            color = color + reflected_color * record.material.reflectance
        
        # Diffuse Beleuchtung (vereinfacht)
        if record.material.color.length() > 0:
            # Einfache Ambient-Beleuchtung
            ambient = record.material.color * 0.3
            
            # Direkte Beleuchtung von Lichtquellen simulieren
            # In einer echten Cornell-Box würden hier mehrere Lichtquellen berechnet
            light_dir = Vec3(0.5, 1, 0.5).normalize()
            light_intensity = max(0, record.normal.dot(light_dir))
            diffuse = record.material.color * light_intensity * 0.7
            
            color = color + ambient + diffuse
        
        return color
    
    def reflect(self, direction: Vec3, normal: Vec3) -> Vec3:
        """Berechnet reflektierten Vektor"""
        return direction - normal * (2 * direction.dot(normal))

# ============================================================================
# Bildausgabe-Modul
# ============================================================================

class ImageWriter:
    """Schreibt ein Bild im PPM-Format"""
    def __init__(self, filename: str, width: int, height: int):
        self.filename = filename
        self.width = width
        self.height = height
        self.pixels = [[Vec3() for _ in range(width)] for _ in range(height)]
    
    def set_pixel(self, x: int, y: int, color: Vec3):
        if 0 <= x < self.width and 0 <= y < self.height:
            self.pixels[y][x] = color
    
    def save(self):
        with open(self.filename, 'w') as f:
            # PPM Header
            f.write(f"P3\n{self.width} {self.height}\n255\n")
            
            for y in range(self.height):
                for x in range(self.width):
                    color = self.pixels[y][x]
                    # Gamma-Korrektur (vereinfacht)
                    r = int(math.sqrt(color.x) * 255)
                    g = int(math.sqrt(color.y) * 255)
                    b = int(math.sqrt(color.z) * 255)
                    
                    # Begrenzung auf [0, 255]
                    r = max(0, min(255, r))
                    g = max(0, min(255, g))
                    b = max(0, min(255, b))
                    
                    f.write(f"{r} {g} {b} ")
                f.write("\n")

# ============================================================================
# Hauptprogramm - Erstellung der Cornell-Box
# ============================================================================

def create_cornell_box() -> World:
    """Erstellt eine Cornell-Box mit verschiedenen Materialien"""
    world = World()
    
    # Materialien definieren
    red = Material(Vec3(1, 0.2, 0.2))          # Rote Wand
    green = Material(Vec3(0.2, 1, 0.2))        # Grüne Wand
    white = Material(Vec3(0.8, 0.8, 0.8))      # Weiße Wände
    light = Material(Vec3(1, 1, 1), Vec3(1, 1, 1) * 2)  # Lichtquelle
    mirror = Material(Vec3(1, 1, 1), reflectance=0.9, roughness=0.1)  # Spiegel
    blue_sphere = Material(Vec3(0.2, 0.2, 1), reflectance=0.3, roughness=0.2)  # Blaue Kugel
    
    # Box-Wände
    # Boden (y=0)
    world.add(Plane(Vec3(0, 0, 0), Vec3(0, 1, 0), white))
    # Decke (y=2)
    world.add(Plane(Vec3(0, 2, 0), Vec3(0, -1, 0), white))
    # Rückwand (z=-2)
    world.add(Plane(Vec3(0, 1, -2), Vec3(0, 0, 1), white))
    # Linke Wand (x=-2) - ROT
    world.add(Plane(Vec3(-2, 1, -1), Vec3(1, 0, 0), red))
    # Rechte Wand (x=2) - GRÜN
    world.add(Plane(Vec3(2, 1, -1), Vec3(-1, 0, 0), green))
    
    # Licht an der Decke
    world.add(Sphere(Vec3(0, 1.8, -0.5), 0.3, light))
    
    # Objekte in der Box
    # Spiegelnde Kugel
    world.add(Sphere(Vec3(-1, 0.5, 0), 0.5, mirror))
    # Blaue Kugel
    world.add(Sphere(Vec3(1, 0.4, -0.5), 0.4, blue_sphere))
    # Kleine weiße Kugel
    world.add(Sphere(Vec3(0, 0.3, 1), 0.3, white))
    
    return world

def main():
    """Hauptfunktion - führt den Raytracer aus und speichert das Bild"""
    # Bildparameter
    width = 800
    height = 600
    aspect = width / height
    
    # Welt erstellen
    world = create_cornell_box()
    
    # Kamera positionieren
    camera = Camera(
        lookfrom=Vec3(0, 1.5, 5),
        lookat=Vec3(0, 1, -1),
        vup=Vec3(0, 1, 0),
        vfov=60,
        aspect=aspect
    )
    
    # Raytracer initialisieren
    raytracer = Raytracer(world, max_depth=5)
    
    # Bild erstellen
    image = ImageWriter("cornell_box.ppm", width, height)
    
    print("Raytracing gestartet...")
    samples_per_pixel = 4  # Für bessere Qualität erhöhen
    
    for y in range(height):
        if y % 50 == 0:
            print(f"Zeile {y}/{height} fertig")
        
        for x in range(width):
            color = Vec3()
            
            # Mehrere Samples pro Pixel für Anti-Aliasing
            for _ in range(samples_per_pixel):
                u = (x + random.random()) / width
                v = (y + random.random()) / height
                
                ray = camera.get_ray(u, v)
                color = color + raytracer.trace(ray, 0)
            
            # Durchschnitt der Samples
            color = color / samples_per_pixel
            image.set_pixel(x, y, color)
    
    # Bild speichern
    image.save()
    print(f"Bild gespeichert als 'cornell_box.ppm'")

if __name__ == "__main__":
    main()

File: Initial/SolutionsByGPT/C++ Leaning/V1Code.cpp
#include <iostream>
#include <fstream>
#include <cmath>
#include <vector>
#include <memory>
#include <limits>

// ===================== Vektor- und Mathematik-Klassen =====================
struct Vec3 {
    double x, y, z;
    Vec3(double x_=0, double y_=0, double z_=0) : x(x_), y(y_), z(z_) {}
    Vec3 operator+(const Vec3 &v) const { return Vec3(x+v.x, y+v.y, z+v.z); }
    Vec3 operator-(const Vec3 &v) const { return Vec3(x-v.x, y-v.y, z-v.z); }
    Vec3 operator*(double s) const { return Vec3(x*s, y*s, z*s); }
    Vec3 operator/(double s) const { return Vec3(x/s, y/s, z/s); }
    Vec3 normalized() const { double len = std::sqrt(x*x + y*y + z*z); return *this / len; }
    double dot(const Vec3 &v) const { return x*v.x + y*v.y + z*v.z; }
    Vec3 cross(const Vec3 &v) const {
        return Vec3(y*v.z - z*v.y, z*v.x - x*v.z, x*v.y - y*v.x);
    }
};

using Color = Vec3;

// ===================== Ray =====================
struct Ray {
    Vec3 origin, direction;
    Ray(const Vec3 &o, const Vec3 &d) : origin(o), direction(d.normalized()) {}
};

// ===================== Material =====================
struct Material {
    Color color;
    double reflectivity;
    Material(const Color &c, double r=0) : color(c), reflectivity(r) {}
};

// ===================== Hitable Primitives =====================
struct HitRecord {
    Vec3 point;
    Vec3 normal;
    std::shared_ptr<Material> material;
    double t;
};

class Hitable {
public:
    virtual bool hit(const Ray &r, double t_min, double t_max, HitRecord &rec) const = 0;
};

// ---------------- Sphere ----------------
class Sphere : public Hitable {
public:
    Vec3 center;
    double radius;
    std::shared_ptr<Material> material;
    Sphere(const Vec3 &c, double r, std::shared_ptr<Material> m) : center(c), radius(r), material(m) {}
    bool hit(const Ray &r, double t_min, double t_max, HitRecord &rec) const override {
        Vec3 oc = r.origin - center;
        double a = r.direction.dot(r.direction);
        double b = oc.dot(r.direction);
        double c = oc.dot(oc) - radius*radius;
        double discriminant = b*b - a*c;
        if (discriminant > 0) {
            double temp = (-b - std::sqrt(discriminant)) / a;
            if(temp < t_max && temp > t_min) {
                rec.t = temp;
                rec.point = r.origin + r.direction*rec.t;
                rec.normal = (rec.point - center) / radius;
                rec.material = material;
                return true;
            }
            temp = (-b + std::sqrt(discriminant)) / a;
            if(temp < t_max && temp > t_min) {
                rec.t = temp;
                rec.point = r.origin + r.direction*rec.t;
                rec.normal = (rec.point - center) / radius;
                rec.material = material;
                return true;
            }
        }
        return false;
    }
};

// ---------------- Plane ----------------
class Plane : public Hitable {
public:
    Vec3 point, normal;
    std::shared_ptr<Material> material;
    Plane(const Vec3 &p, const Vec3 &n, std::shared_ptr<Material> m) : point(p), normal(n.normalized()), material(m) {}
    bool hit(const Ray &r, double t_min, double t_max, HitRecord &rec) const override {
        double denom = normal.dot(r.direction);
        if(std::abs(denom) > 1e-6) {
            double t = (point - r.origin).dot(normal) / denom;
            if(t >= t_min && t <= t_max) {
                rec.t = t;
                rec.point = r.origin + r.direction*t;
                rec.normal = normal;
                rec.material = material;
                return true;
            }
        }
        return false;
    }
};

// ===================== Scene =====================
class Scene {
public:
    std::vector<std::shared_ptr<Hitable>> objects;
    void add(std::shared_ptr<Hitable> obj) { objects.push_back(obj); }

    bool hit(const Ray &r, double t_min, double t_max, HitRecord &rec) const {
        HitRecord tempRec;
        bool hitAnything = false;
        double closestSoFar = t_max;
        for(const auto &obj : objects) {
            if(obj->hit(r, t_min, closestSoFar, tempRec)) {
                hitAnything = true;
                closestSoFar = tempRec.t;
                rec = tempRec;
            }
        }
        return hitAnything;
    }
};

// ===================== Raytracer =====================
Color rayColor(const Ray &r, const Scene &scene, int depth) {
    if(depth <= 0) return Color(0,0,0);
    HitRecord rec;
    if(scene.hit(r, 0.001, std::numeric_limits<double>::max(), rec)) {
        Vec3 target = rec.point + rec.normal;
        Ray reflectedRay(rec.point, r.direction - rec.normal*2*r.direction.dot(rec.normal));
        Color reflectedColor = rayColor(reflectedRay, scene, depth-1);
        return rec.material->color*(1-reflectedColor.dot(Color(1,1,1))*rec.material->reflectivity) + reflectedColor*rec.material->reflectivity;
    }
    // Hintergrundfarbe
    Vec3 unit_direction = r.direction.normalized();
    double t = 0.5*(unit_direction.y + 1.0);
    return Color(1.0,1.0,1.0)*(1.0-t) + Color(0.5,0.7,1.0)*t;
}

// ===================== Main =====================
int main() {
    // Bildgröße
    const int imageWidth = 400;
    const int imageHeight = 400;
    const int maxDepth = 5;

    // Kamera
    Vec3 origin(0,1,3);
    double viewportHeight = 2.0;
    double viewportWidth = 2.0;
    double focalLength = 1.0;
    Vec3 horizontal(viewportWidth,0,0);
    Vec3 vertical(0,viewportHeight,0);
    Vec3 lowerLeftCorner = origin - horizontal/2 - vertical/2 - Vec3(0,0,focalLength);

    // Szene
    Scene scene;
    auto red   = std::make_shared<Material>(Color(0.75,0.1,0.1));
    auto green = std::make_shared<Material>(Color(0.1,0.75,0.1));
    auto white = std::make_shared<Material>(Color(0.75,0.75,0.75));
    auto mirror= std::make_shared<Material>(Color(0.9,0.9,0.9), 0.7);

    // Cornellbox Wände
    scene.add(std::make_shared<Plane>(Vec3(0,0,0), Vec3(0,1,0), white)); // Boden
    scene.add(std::make_shared<Plane>(Vec3(0,2,0), Vec3(0,-1,0), white)); // Decke
    scene.add(std::make_shared<Plane>(Vec3(-1,0,0), Vec3(1,0,0), red)); // links
    scene.add(std::make_shared<Plane>(Vec3(1,0,0), Vec3(-1,0,0), green)); // rechts
    scene.add(std::make_shared<Plane>(Vec3(0,0,-1), Vec3(0,0,1), white)); // hinten

    // Kugel in der Box
    scene.add(std::make_shared<Sphere>(Vec3(-0.4,0.35,-0.3), 0.35, mirror));
    scene.add(std::make_shared<Sphere>(Vec3(0.5,0.3,-0.6), 0.3, white));

    // Bilddatei
    std::ofstream image("cornell.ppm");
    image << "P3\n" << imageWidth << " " << imageHeight << "\n255\n";

    for(int j = imageHeight-1; j >= 0; --j) {
        for(int i = 0; i < imageWidth; ++i) {
            double u = double(i)/ (imageWidth-1);
            double v = double(j)/ (imageHeight-1);
            Ray r(origin, lowerLeftCorner + horizontal*u + vertical*v - origin);
            Color col = rayColor(r, scene, maxDepth);
            int ir = static_cast<int>(255.999 * std::fmin(1.0, col.x));
            int ig = static_cast<int>(255.999 * std::fmin(1.0, col.y));
            int ib = static_cast<int>(255.999 * std::fmin(1.0, col.z));
            image << ir << " " << ig << " " << ib << "\n";
        }
    }

    image.close();
    std::cout << "Bild 'cornell.ppm' erstellt.\n";
    return 0;
}
